clients/src/main/java/org/apache/kafka/common/requests/CreateTopicRequest.java

```diff
@@ -0,0 +1,212 @@
+        }
+
+        public TopicDetails(Map<Integer, List<Integer>> replicasAssignments) {
+            this(replicasAssignments, new HashMap<String, String>());
+        }
+    }
+
+    private final Map<String, TopicDetails> topics;
+    private final Integer timeout;
+
+    public static final int NO_PARTITIONS_SIGN = -1;
+    public static final int NO_REPLICATION_FACTOR_SIGN = -1;
+
+    public CreateTopicRequest(Map<String, TopicDetails> topics, Integer timeout) {
+        super(new Struct(CURRENT_SCHEMA));
+
+        List<Struct> createTopicRequestStructs = new ArrayList<>(topics.size());
+        for (Map.Entry<String, TopicDetails> entry : topics.entrySet()) {
+
+            Struct singleRequestStruct = struct.instance(REQUESTS_KEY_NAME);
+            String topic = entry.getKey();
+            TopicDetails args = entry.getValue();
+
+            singleRequestStruct.set(TOPIC_KEY_NAME, topic);
+            singleRequestStruct.set(PARTITIONS_KEY_NAME, args.partitions);
+            singleRequestStruct.set(REPLICATION_FACTOR_KEY_NAME, args.replicationFactor);
+
+            // replica assignment
+            List<Struct> replicaAssignmentsStructs = new ArrayList<>(args.replicasAssignments.size());
+            for (Map.Entry<Integer, List<Integer>> partitionReplicaAssignment : args.replicasAssignments.entrySet()) {
+                Struct replicaAssignmentStruct = singleRequestStruct.instance(REPLICA_ASSIGNMENT_KEY_NAME);
+                replicaAssignmentStruct.set(REPLICA_ASSIGNMENT_PARTITION_ID_KEY_NAME, partitionReplicaAssignment.getKey());
+                replicaAssignmentStruct.set(REPLICA_ASSIGNMENT_REPLICAS_KEY_NAME, partitionReplicaAssignment.getValue().toArray());
+                replicaAssignmentsStructs.add(replicaAssignmentStruct);
+            }
+            singleRequestStruct.set(REPLICA_ASSIGNMENT_KEY_NAME, replicaAssignmentsStructs.toArray());
+
+            // configs
+            List<Struct> configsStructs = new ArrayList<>(args.configs.size());
+            for (Map.Entry<String, String> configEntry : args.configs.entrySet()) {
+                Struct configStruct = singleRequestStruct.instance(CONFIGS_KEY_NAME);
+                configStruct.set(CONFIG_KEY_KEY_NAME, configEntry.getKey());
+                configStruct.set(CONFIG_VALUE_KEY_NAME, configEntry.getValue());
+                configsStructs.add(configStruct);
+            }
+            singleRequestStruct.set(CONFIGS_KEY_NAME, configsStructs.toArray());
+            createTopicRequestStructs.add(singleRequestStruct);
+        }
+        struct.set(REQUESTS_KEY_NAME, createTopicRequestStructs.toArray());
+        struct.set(TIMEOUT_KEY_NAME, timeout);
+
+        this.topics = topics;
+        this.timeout = timeout;
+    }
+
+    public CreateTopicRequest(Struct struct) {
+        super(struct);
+
+        Object[] requestStructs = struct.getArray(REQUESTS_KEY_NAME);
+        topics = new HashMap<>();
```

Comment:
To have the right concurrency semantics, you should not mutate the map after assigning it to the final field.